### PR TITLE
Stabilize `param_attrs` in Rust 1.39.0

### PR DESCRIPTION
**Pull request by c410-f3r**
# Stabilization proposal

I propose that we stabilize `#![feature(param_attrs)]`.

Tracking issue: #60406
Version: 1.39 (2019-09-26 => beta, 2019-11-07 => stable).

## What is stabilized

It is now possible to add outer attributes like `#[cfg(..)]` on formal parameters of functions, closures, and function pointer types. For example:

```rust
fn len(
    #[cfg(windows)] slice: &[u16],
    #[cfg(not(windows))] slice: &[u8],
) -> usize {
    slice.len()
}
```

## What isn't stabilized

* Documentation comments like `/// Doc` on parameters.

* Code expansion of a user-defined `#[proc_macro_attribute]` macro used on parameters.

* Built-in attributes other than `cfg`, `cfg_attr`, `allow`, `warn`, `deny`, and `forbid`. Currently, only the lints `unused_variables` and `unused_mut` have effect and may be controlled on parameters.

## Motivation

The chief motivations for stabilizing `param_attrs` include:

* Finer conditional compilation with `#[cfg(..)]` and linting control of variables.

* Richer macro DSLs created by users.

* External tools and compiler internals can take advantage of the additional information that the parameters provide.

For more examples, see the [RFC][rfc motivation].

## Reference guide

In the grammar of function and function pointer, the grammar of variadic tails (`...`) and parameters are changed respectively from:

```rust
FnParam = { pat:Pat ":" }? ty:Type;
VaradicTail = "...";
```

into:

```rust
FnParam = OuterAttr* { pat:Pat ":" }? ty:Type;
VaradicTail = OuterAttr* "...";
```

The grammar of a closure parameter is changed from:

```rust
ClosureParam = pat:Pat { ":" ty:Type }?;
```

into:

```rust
ClosureParam = OuterAttr* pat:Pat { ":" ty:Type }?;
```

More generally, where there's a list of formal (value) parameters separated or terminated by `,` and delimited by `(` and `)`. Each parameter in that list may optionally be prefixed by `OuterAttr+`.

Note that in all cases, `OuterAttr*` applies to the whole parameter and not just the pattern. This distinction matters in pretty printing and in turn for macros.

## History

* On 2018-10-15, @Robbepop proposes [RFC 2565][rfc], "Attributes in formal function parameter position".

* On 2019-04-30, [RFC 2565][rfc] is merged and the tracking issue is made.

* On 2019-06-12, a partial implementation was completed. The implementation was done in [#60669][60669] by @c410-f3r and the PR was reviewed by @petrochenkov and @Centril.

* On 2019-07-29, [#61238][61238] was fixed in [#61856][61856]. The issue fixed was that lint attributes on function args had no effect. The PR was written by @c410-f3r and reviewed by @matthewjasper, @petrochenkov, and @oli-obk.

* On 2019-08-02, a bug [#63210][63210] was filed wherein the attributes on formal parameters would not be passed to macros. The issue was about forgetting to call the relevant method in `fn print_arg` in the pretty printer. In [#63212][63212], written by @Centril on 2019-08-02 and reviewed by @davidtwco, the issue aforementioned was fixed.

* This PR stabilizes `param_attrs`.

## Tests

* [On Rust 2018, attributes aren't permitted on function parameters without a pattern in trait definitions.](https://github.com/rust-lang/rust/blob/master/src/test/ui/rfc-2565-param-attrs/param-attrs-2018.rs)

* [All attributes that should be allowed. This includes `cfg`, `cfg_attr`, and lints check attributes.](https://github.com/rust-lang/rust/blob/master/src/test/ui/rfc-2565-param-attrs/param-attrs-allowed.rs)

* [Built-in attributes, which should be forbidden, e.g., `#[test]`, are.](https://github.com/rust-lang/rust/blob/master/src/test/ui/rfc-2565-param-attrs/param-attrs-builtin-attrs.rs)

* [`cfg` and `cfg_attr` are properly evaluated.](https://github.com/rust-lang/rust/blob/master/src/test/ui/rfc-2565-param-attrs/param-attrs-cfg.rs)

* [`unused_mut`](https://github.com/rust-lang/rust/blob/46f405ec4d7c6bf16fc2eaafe7541019f1da2996/src/test/ui/rfc-2565-param-attrs/param-attrs-cfg.rs) and [`unused_variables`](https://github.com/rust-lang/rust/blob/master/src/test/ui/lint/lint-unused-variables.rs) are correctly applied to parameter patterns.

* [Pretty printing takes formal parameter attributes into account.](https://github.com/rust-lang/rust/blob/master/src/test/ui/rfc-2565-param-attrs/param-attrs-pretty.rs)

## Possible future work

* Custom attributes inside function parameters aren't currently supported but it is something being worked on internally.

* Since documentation comments are syntactic sugar for `#[doc(...)]`, it is possible to allow literal `/// Foo` comments on function parameters.

[rfc motivation]: https://github.com/rust-lang/rfcs/blob/master/text/2565-formal-function-parameter-attributes.md#motivation
[rfc]: https://github.com/rust-lang/rfcs/pull/2565
[60669]: https://github.com/rust-lang/rust/pull/60669
[61856]: https://github.com/rust-lang/rust/pull/61856
[63210]: https://github.com/rust-lang/rust/issues/63210
[61238]: https://github.com/rust-lang/rust/issues/61238
[63212]: https://github.com/rust-lang/rust/pull/63212

This report is a collaborative work with @Centril.

### CODE REVIEW

File: src/test/ui/rfc-2565-param-attrs/param-attrs-feature-gate.rs [hunk 1]
```diff
@@ -6,11 +6,7 @@ fn foo(
     /// Foo
```
Comment (Centril) [inline src/test/ui/rfc-2565-param-attrs/param-attrs-feature-gate.rs]: This file and the associated `.stderr` file should be removed.

**Comment by Centril** [comment]
Dear language team and the community at large. The above report gives an overview of the `#[attr] param: type` feature. I propose that we indeed do stabilize the feature accordingly.

@rfcbot merge

**Review by Centril — APPROVED** [review]
r=me on the code changes itself when FCP completes.

**Comment by phansch** [comment]
> Currently, only the lints unused_variables and unused_mut have effect, and may be controlled, on parameters.

How does that work? Was it a deliberate decision somewhere in the code or a coincidental result of the implementation? I'm just asking to find out if there's something we should be aware of with tool lints in Clippy.

**Comment by nikomatsakis** [comment]
Question for @Centril and @c410-f3r -- 

You mention that procedural macros are not stabilized. Is there a test showing what happens when a procedural macro is applied? I presume that it generates a feature gate? Or is it considered an error?

**Comment by nikomatsakis** [comment]
(Also, I should add, thanks for all your work on this. ❤️ )

**Comment by c410-f3r** [comment]
@phansch 

According to the [RFC](https://github.com/rust-lang/rfcs/blob/master/text/2565-formal-function-parameter-attributes.md#built-in-attributes), this reduced number of allowed lint was initially proposed to be part of a minimum viable product. Nevertheless, more lints can be added in future releases.

@nikomatsakis 

Currently, the compiler shows the `the attribute 'some_attribute' is currently unknown to the compiler and may have meaning added to it in the future` error. With #63468, it will be changed to `expected an inert attribute, found an attribute macro` and the message itself [behaves as a feature gate](https://github.com/rust-lang/rust/pull/63468#issuecomment-520437663).

**Comment by Centril** [comment]
@nikomatsakis 

> You mention that procedural macros are not stabilized. Is there a test showing what happens when a procedural macro is applied? I presume that it generates a feature gate? Or is it considered an error?

https://github.com/rust-lang/rust/blob/master/src/test/ui/rfc-2565-param-attrs/param-attrs-builtin-attrs.rs sorta addresses this by using `#[test]` but I've hardened the tests somewhat in https://github.com/rust-lang/rust/pull/64031.

> Similarly, is there a test for lints that are intended to be disallowed?

The RFC does not specify a specific list of lints that should or should not be allowed but the two lints that are tested and accounted for seemed the most important. We'd need to do a deeper audit of all the lints to see which others could be relevant and that's probably not worth the time at the moment.

**Comment by Centril** [comment]
> How does that work? Was it a deliberate decision somewhere in the code or a coincidental result of the implementation? I'm just asking to find out if there's something we should be aware of with tool lints in Clippy.

@phansch Reading https://github.com/rust-lang/rust/pull/61856 is probably most relevant for Clippy development.

**Comment by scottmcm** [comment]
Just to confirm, the "applies to the whole parameter and not just the pattern" means that this works even for things like `fn foo(#[bar] (&x, _):(&i32, i32))`, right?  And removes the whole parameter?

**Comment by Centril** [comment]
@scottmcm Yes, correct. The rough grammar is `OuterAttr* Pattern ":" Type`. For details, see: https://github.com/rust-lang/rust/blob/59cc53e6e708e9b9e86822c1b4d69f28f6c45eae/src/libsyntax/parse/parser.rs#L982

(That code probably needs some cleaning up to make the happy path clearer...)

**Comment by nikomatsakis** [comment]
@Centril 

> The RFC does not specify a specific list of lints that should or should not be allowed but the two lints that are tested and accounted for seemed the most important. We'd need to do a deeper audit of all the lints to see which others could be relevant and that's probably not worth the time at the moment.

Maybe I'm confused. The wording from the report suggested to me that `#[allow]` on a parameter was meant to error or give a warning or something like that?

**Comment by Centril** [comment]
@nikomatsakis 

> Maybe I'm confused. The wording from the report suggested to me that `#[allow]` on a parameter was meant to error or give a warning or something like that, when used with an incorrect line?

[The RFC specifies](https://github.com/rust-lang/rfcs/blob/master/text/2565-formal-function-parameter-attributes.md#built-in-attributes-1) and [we have a test that ensures](https://github.com/rust-lang/rust/blob/master/src/test/ui/rfc-2565-param-attrs/param-attrs-allowed.rs) that `fn foo(#[allow/warn/deny/forbid(lint_name)] bar: Baz)` is legal. If `#[allow(foo)]` is used on a parameter and the lint `foo` exists but will never apply in that context, then nothing happens.

The normal operation of e.g., `#[allow(lint_name)]` is that
1. [an unknown lint, e.g. `#[allow(foo)]` will generate a warning.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2df8aa5b02b87e38ba6dfd1687255d3c)
2. it only suppresses lints that are actually fired in the context the attribute is put on. `#[allow(lint_name)]` has no effect if no such lint is emitted in that context. For example, in this snippet...
   ```rust
   #[allow(unused_must_use)]
   pub struct Foo(u8);
   ```
   ...the `#[allow(..)]` has no effect as the lint will not fire, and no warning is emitted saying it cannot fire.

>  i.e., one we had not whitelisted.

To my knowledge we do not whitelist specific lints in specific contexts. Rather, a lint will use [`cx.struct_span_lint(...)`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/lint/context/trait.LintContext.html#method.struct_span_lint) and behind the scenes, the lint infrastructure will determine the current `Level` and eventually call [`lint::struct_lint_level`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/lint/fn.struct_lint_level.html) which conditionally emits the lint.

What https://github.com/rust-lang/rust/pull/61856 did was to adjust the visitors so that they collect `#[allow(...)]` and friends on parameters as well as to introduce [`fn check_param`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/lint/trait.EarlyLintPass.html#method.check_param). Some logic is also introduced to make the MIR lints trigger in the relevant context: https://github.com/rust-lang/rust/pull/61856/files#diff-2daa9fd281d82f3fc811f6058691747bR878.

> Are you assuming instead that we'll simply grow the set of lints that apply as needed? (That does seem like it'd be within our stability policy, so perhaps just fine)

Yep. I believe that's what we usually do. I think it's clearly within our stability policy to tweak where lints apply and which ones do.

**Comment by Centril** [comment]
https://github.com/rust-lang/rust/pull/64031 just landed, assuring that https://github.com/rust-lang/rust/pull/64010#issuecomment-526564316 is addressed with more tests.

@c410-f3r Can you rebase and remove the feature gate in that new added test as well?

**Review by Centril — APPROVED** [review]
Same as before... r=me when FCP completes ;)

**Comment by Centril** [comment]
@rfcbot concern #64282

**Comment by Centril** [comment]
@rfcbot resolve #64282

**Comment by pnkfelix** [comment]
> Question for @Centril and @c410-f3r --
> 
> You mention that procedural macros are not stabilized. Is there a test showing what happens when a procedural macro is applied? I presume that it generates a feature gate? Or is it considered an error?
> 
> Similarly, is there a test for lints that are intended to be disallowed?

Hmm. I wonder if its would be feasible (and worthwhile), in the future, for descriptions in these stabilization reports to actually link to the corresponding tests when there are specific comments like this. (It isn't always something in the diff, e.g. when the test arises from some code that was already in the repository.)

Of course it would make authoring such reports even more arduous. But it seems like it might streamline the review process...

**Comment by yanfenglee** [comment]
When Custom attributes inside function parameters available? Is there a plan?

**Comment by c410-f3r** [comment]
It is currently possible to parse custom attributes using procedural macros but they can't be present in the final `TokenStream` output.

**Comment by yanfenglee** [comment]
> It is currently possible to parse custom attributes using procedural macros but they can't be present in the final `TokenStream` output.

Thanks for your reply, i don't quite understand "they can't be present in the final `TokenStream` output.",  i want implement like this:
```rust
#[post("/path/{id}")]
async fn post_data(data: &PostData, #[path_varible("id")]id: i32, #[header("token")] token: String) -> Result<String, Error> {}
```
Is this possible?

**Comment by c410-f3r** [comment]
Taking into consideration that `#[post]` is a procedural macro created by you, then yes. It is possible to parse `#[path_varible("id")]` and  `#[header("token")]`.

> Thanks for your reply, i don't quite understand "they can't be present in the final TokenStream output.", i want implement like this:

`#[post]` is a procedural attribute macro that has the following signature: `fn post(attrs: TokenStream, item: TokenStream) -> TokenStream;`. Although `[path_varible("id")]` and `#[header("token")]` are present in `item`, they shouldn't be present in the `TokenStream` output.

Take a look at https://doc.rust-lang.org/reference/procedural-macros.html to learn more about procedural macros, use the `syn` crate for parsing and the `quote` for building tokens.

**Pull request merged**